lib/E2E/LowerRankedShapes.cpp

```diff
@@ -193,14 +247,18 @@ class LowerRankedShapes : public LowerRankedShapesBase<LowerRankedShapes> {
     // Erase some stray shape ops from the program. They can't be
     // deleted during conversion because they become unused only after
     // subsequent patterns bypass them.
+    // TODO: This kind of DCE plus verification in one step is too simplistic
```

Comment:
I don't understand how it constrains programs more than desired. If the pass cannot establish its invariants, the pipeline will fail mysteriously elsewhere with some dangling op floating around that still uses !shape.shape or an extent tensor.

Reply:
I'll rephrase it: I spent a fair amount of time trying to make this do the right level of DCE and enforcement and failed. I'll rephrase the TODO to mirror the one where I inserted the canonicalizer pass.